types/parser/visitor.go

```diff
@@ -0,0 +1,319 @@
+	nullability := types.NullabilityRequired
+	if ctx.GetIsnull() != nil {
+		nullability = types.NullabilityNullable
+	}
+
+	precision := v.Visit(ctx.GetPrecision()).(integer_parameters.IntegerParameter)
+	scale := v.Visit(ctx.GetScale()).(integer_parameters.IntegerParameter)
+
+	return &types.ParameterizedDecimalType{Precision: precision, Scale: scale, Nullability: nullability}
+}
+
+func (v *TypeVisitor) VisitPrecisionTimestamp(ctx *baseparser2.PrecisionTimestampContext) interface{} {
+	nullability := types.NullabilityRequired
+	if ctx.GetIsnull() != nil {
+		nullability = types.NullabilityNullable
+	}
+
+	length := v.Visit(ctx.GetPrecision()).(integer_parameters.IntegerParameter)
+	return &types.ParameterizedPrecisionTimestampType{IntegerOption: length, Nullability: nullability}
+}
+
+func (v *TypeVisitor) VisitPrecisionTimestampTZ(ctx *baseparser2.PrecisionTimestampTZContext) interface{} {
+	nullability := types.NullabilityRequired
+	if ctx.GetIsnull() != nil {
+		nullability = types.NullabilityNullable
+	}
+
+	length := v.Visit(ctx.GetPrecision()).(integer_parameters.IntegerParameter)
+	return &types.ParameterizedPrecisionTimestampTzType{IntegerOption: length, Nullability: nullability}
+}
+
+func (v *TypeVisitor) VisitPrecisionIntervalDay(ctx *baseparser2.PrecisionIntervalDayContext) interface{} {
+	nullability := types.NullabilityRequired
+	if ctx.GetIsnull() != nil {
+		nullability = types.NullabilityNullable
+	}
+
+	length := v.Visit(ctx.GetPrecision()).(integer_parameters.IntegerParameter)
+	return &types.ParameterizedIntervalDayType{IntegerOption: length, Nullability: nullability}
+}
+
+func (v *TypeVisitor) VisitStruct(ctx *baseparser2.StructContext) interface{} {
+	nullability := types.NullabilityRequired
+	if ctx.GetIsnull() != nil {
+		nullability = types.NullabilityNullable
+	}
+	var fieldTypes []types.FuncDefArgType
+	for _, expr := range ctx.AllExpr() {
+		fieldTypes = append(fieldTypes, v.Visit(expr).(types.FuncDefArgType))
+	}
+	return &types.ParameterizedStructType{Types: fieldTypes, Nullability: nullability}
+}
+
+func (v *TypeVisitor) VisitList(ctx *baseparser2.ListContext) interface{} {
+	nullability := types.NullabilityRequired
+	if ctx.GetIsnull() != nil {
+		nullability = types.NullabilityNullable
+	}
+	elementType := v.Visit(ctx.Expr()).(types.FuncDefArgType)
+	return &types.ParameterizedListType{Type: elementType, Nullability: nullability}
+}
+
+func (v *TypeVisitor) VisitNStruct(*baseparser2.NStructContext) interface{} {
+	panic("implement me")
+}
+
+func (v *TypeVisitor) VisitMap(ctx *baseparser2.MapContext) interface{} {
+	nullability := types.NullabilityRequired
+	if ctx.GetIsnull() != nil {
+		nullability = types.NullabilityNullable
+	}
+	keyType, keyOk := v.Visit(ctx.GetKey()).(types.FuncDefArgType)
+	valueType, valueOk := v.Visit(ctx.GetValue()).(types.FuncDefArgType)
+	if !keyOk || !valueOk {
+		panic("Map key or value type is not a FuncDefArgType")
```

Comment:
Do we want production code to have to deal with panics?  An alternative would be to use antlr's error handling mechanism.